### PR TITLE
refs #53 handle invalid content-type headers

### PR DESCRIPTION
**Pull request by gingerlime**
* added a test to simulate the issue
* unfortunately, not entirely sure how to catch this exception and
  handle it yet ... so this test currently fails

### CODE REVIEW

**Comment by yuki24** [comment]
You could probably sanitize the `Content-Type` header by doing:

```ruby
begin
  request.content_type
rescue Mime::Type::InvalidMimeType
  request.env["MALFORMED_HTTP_CONTENT_TYPE"], request.env["HTTP_CONTENT_TYPET"] = request.env["HTTP_CONTENT_TYPE"], "*/*"
end
```

And you can probably add it to either before or after where the `Accept` header is sannitized: https://github.com/yuki24/rambulance/blob/75e4d4a22d90b2b950b256bb73b4b50bbbe898ec/lib/rambulance/exceptions_app.rb#L70-L74

**Comment by gingerlime** [comment]
Thank @yuki24 ! I tried something similar. The problem I see is that `request.POST` then fails, even if I update the request content_type beforehand... I'll push another commit that fixes it, but I'm not entirely sure if it's the best solution still (at least you can hopefully see what I mean)

File: test/requests/error_json_test.rb [hunk 1]
```diff
@@ -67,12 +93,12 @@ def without_layouts
     `mv error.html.erb test/fake_app/app/views/layouts/`
   end
 
-  def get(path)
+  def get(path, headers={})
```
Comment (yuki24) [inline test/requests/error_json_test.rb:96]:
```suggestion
  def get(path, headers: {})
```

File: lib/rambulance/exceptions_app.rb [hunk 2]
```diff
@@ -59,12 +59,16 @@ def process_action(*)
         request.GET
       rescue ActionController::BadRequest
         request.env["MALFORMED_QUERY_STRING"], request.env["QUERY_STRING"] = request.env["QUERY_STRING"], ""
+      rescue Mime::Type::InvalidMimeType
+        request.env["MALFORMED_CONTENT_TYPE"], request.env["CONTENT_TYPE"] = request.env["CONTENT_TYPE"], "text/plain"
```
Comment (yuki24) [inline lib/rambulance/exceptions_app.rb:63]: Each one of the elements should be sanitized independently to cover a case where both the query string and the content type header. You could probably do this at the beginning of this method:

```ruby
def process_action(*)
  begin
    request.content_mime_type
  rescue Mime::Type::InvalidMimeType
    request.env["MALFORMED_CONTENT_TYPE"], request.env["CONTENT_TYPE"] = request.env["CONTENT_TYPE"], "*/*"
  end

  ...
```

File: lib/rambulance/exceptions_app.rb [hunk 3]
```diff
@@ -59,12 +59,16 @@ def process_action(*)
...
       end
 
       begin
         request.POST
       rescue ActionController::BadRequest
         request.env["MALFORMED_BODY"], request.env["rack.input"] = request.env["rack.input"], StringIO.new
+      rescue Mime::Type::InvalidMimeType
+        request.env["MALFORMED_CONTENT_TYPE"], request.env["CONTENT_TYPE"] = request.env["CONTENT_TYPE"], "text/plain"
```
Comment (yuki24) [inline lib/rambulance/exceptions_app.rb:71]: Let's remove this in favor of https://github.com/yuki24/rambulance/pull/55/files#r431494203.

File: test/requests/error_json_test.rb [hunk 4]
```diff
@@ -55,6 +55,32 @@ class ErrorJsonTest < ActionDispatch::IntegrationTest
...
+      assert_equal 201, response.status
+    end
+  end
+
+  test 'GET returns an appropriate status based on the rails version when the HTTP Content-type header is malformed' do
+    get '/projects.json', headers: { "CONTENT_TYPE" => "charset=gbk" }
+
+    assert_equal 406, response.status
+    assert_equal "The requested content type is not acceptable.\n", response.body
+  end
```
Comment (yuki24) [inline test/requests/error_json_test.rb:82]: Ditto https://github.com/yuki24/rambulance/pull/55/files#r431494754

File: test/requests/error_json_test.rb [hunk 5]
```diff
@@ -67,12 +93,12 @@ def without_layouts
     `mv error.html.erb test/fake_app/app/views/layouts/`
   end
 
-  def get(path)
+  def get(path, headers: {})
```
Comment (yuki24) [inline test/requests/error_json_test.rb:96]: I think this is a good change so the signature will be more aligned with , but it would be great if you could send a separate pull request for that.

```ruby
def get(path, params: {}, headers: {})
  ...
end
```

File: lib/rambulance/exceptions_app.rb (hunk 2 above)
Comment (gingerlime) [reply lib/rambulance/exceptions_app.rb:63]: cool. this seems to work! I'll update it. Thanks @yuki24

File: test/requests/error_json_test.rb (hunk 5 above)
Comment (gingerlime) [reply test/requests/error_json_test.rb:96]: That's a good idea. Although I do need it in order to have more flexibility testing with GET since you overwrote it. I'll create a separate PR, and then see if I can rebase this PR off the other one.

Comment (gingerlime) [reply test/requests/error_json_test.rb:96]: see #56 (merged onto this PR)

**Comment by yuki24** [comment]
The test failure seems to be caused by the extra test which I think we can remove. I'll just go ahead nad merge since this pull request looks good as it is.

**Comment by gingerlime** [comment]
That's great. Thanks so much @yuki24 and for your patience helping me work through this.
